File: gates/bohemia_city_src.py
#!/usr/bin/env python3
"""BOHEMIA — WHERE THE CITY LIVES, IN ONE PLACE (8/4/26). Python twin of
gates/bohemia_city_src.js; the full story is in that file's header.

The short version: until 8/4 the walked world was a 35.76 MB base64 constant
inside the alpha (CITY_B64). The CITY lane extracted it to
slices/BOHEMIA_CITY_WORLD.html so the alpha opens 29x faster, and nineteen
gates that hunted for that constant went red at once - for a reason that had
nothing to do with anybody's code. When a third of the suite is red like that,
red stops meaning anything, which is the real damage.

One answer to "where is the city", asked by every gate. Next time the world
moves, one edit here.

    from bohemia_city_src import city_src
    frame = city_src()                 # the city's HTML source, as a string
    frame = city_src(optional=True)    # None instead of raising

Reads only. Owns nothing. Decides nothing.
"""
import base64
import os

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
STANDALONE = os.path.join(ROOT, 'slices', 'BOHEMIA_CITY_WORLD.html')
ALPHA = os.path.join(ROOT, 'slices', 'BOHEMIA_ALPHA_0_9.html')


def _from_alpha(alpha):
    """the old inline form, kept so an older checkout still measures. BY INDEX,
    never a regex: the alpha was 42 MB and a big quantifier over it is a
    stack-blower (learned 7/31)."""
    key = "CITY_B64"
    ci = alpha.find(key)
    while ci >= 0:
        tail = alpha[ci + 8:ci + 20]
        eq = tail.find('=')
        if eq >= 0:
            rest = tail[eq:]
            qi = -1
            for i, ch in enumerate(rest):
                if ch in "'\"`":
                    qi = i
                    break
            if qi >= 0:
                start = ci + 8 + eq + qi + 1
                end = alpha.find(alpha[start - 1], start)
                if end - start >= 100000:
                    return base64.b64decode(alpha[start:end]).decode('utf8', 'replace')
        ci = alpha.find(key, ci + 1)
    return None


def city_src(alpha_text=None, optional=False):
    if os.path.exists(STANDALONE):
        with open(STANDALONE, encoding='utf8', errors='replace') as f:
            s = f.read()
        if len(s) > 100000:
            return s
    alpha = alpha_text
    if alpha is None and os.path.exists(ALPHA):
        with open(ALPHA, encoding='utf8', errors='replace') as f:
            alpha = f.read()
    inline = _from_alpha(alpha) if alpha else None
    if inline:
        return inline
    if optional:
        return None
    raise RuntimeError(
        'the walked city could not be found. Looked for '
        'slices/BOHEMIA_CITY_WORLD.html and for an inline CITY_B64 in the alpha. '
        'If the world moved again, gates/bohemia_city_src.py is the ONE place to '
        'teach it the new address.')


def where():
    return STANDALONE if os.path.exists(STANDALONE) else ALPHA + ' (inline CITY_B64)'

File: gates/bohemia_city_src.js
/* BOHEMIA — WHERE THE CITY LIVES, IN ONE PLACE (8/4/26).

   THE CITY MOVED HOUSE AND TOOK NINETEEN GATES DOWN WITH IT. Until 8/4 the
   walked world was a base64 constant inside the alpha:

       const CITY_B64='PCFET0NUWVBFIGh0bWw+...'      35.76 MB, one line

   The CITY lane extracted it to slices/BOHEMIA_CITY_WORLD.html so the alpha
   opens 29x faster, which is plainly right. But TWENTY-ONE gates read the city
   by hunting for that constant, each with its own hand-rolled extractor copied
   from the last one. Two were migrated with the move. Nineteen were not, and
   they all went red at once - GRAVEYARD, MAP SIZE, STREET SOURCE, FOOTSTEP,
   TRAFFIC SIGNAL, FULL RES, DOORS, INTERIORS, ICON and more.

   THAT IS WORSE THAN IT SOUNDS. "Green or it does not ship" is the law every
   lane works under. When a third of the suite is red for a reason that has
   nothing to do with anybody's code, red stops meaning anything, and the next
   real breakage arrives in a suite nobody is reading any more.

   SO: ONE ANSWER TO "WHERE IS THE CITY", and every gate asks it. The next time
   the world moves - and it will, this is the second home it has had - that is
   ONE edit here instead of nineteen archaeological digs.

   It prefers the standalone file and falls back to the old inline constant, so
   this works on a fresh checkout of main AND on any older tree.

     const citySrc = require('./bohemia_city_src.js');
     const frame = citySrc();          // the city's HTML source, as a string
     const frame = citySrc(null, {optional: true});   // null instead of throwing

   Reads only. Owns nothing. Decides nothing. */
'use strict';
const fs = require('fs');
const path = require('path');

const ROOT = path.dirname(__dirname);
const STANDALONE = path.join(ROOT, 'slices/BOHEMIA_CITY_WORLD.html');
const ALPHA = path.join(ROOT, 'slices/BOHEMIA_ALPHA_0_9.html');

/* the old inline form, kept so an older checkout still measures.
   BY INDEX, NEVER A REGEX: the alpha was 42 MB and a big quantifier over it
   blows the call stack (learned the hard way on 7/31, and every one of the
   nineteen copies of this loop carries that same comment). */
function fromAlpha(alpha) {
  for (let ci = alpha.indexOf('CITY_B64'); ci >= 0; ci = alpha.indexOf('CITY_B64', ci + 1)) {
    const tail = alpha.slice(ci + 8, ci + 20);
    const eq = tail.indexOf('=');
    if (eq < 0) continue;
    const qi = tail.slice(eq).search(/['"`]/);
    if (qi < 0) continue;
    const start = ci + 8 + eq + qi + 1;
    const end = alpha.indexOf(alpha[start - 1], start);
    if (end - start < 100000) continue;
    return Buffer.from(alpha.slice(start, end), 'base64').toString('utf8');
  }
  return null;
}

/* alphaText: pass one if you already read the alpha, else it is read here.
   opts.optional: return null rather than throwing when the city cannot be found. */
function citySrc(alphaText, opts) {
  opts = opts || {};
  if (fs.existsSync(STANDALONE)) {
    const s = fs.readFileSync(STANDALONE, 'utf8');
    if (s.length > 100000) return s;
  }
  const alpha = alphaText || (fs.existsSync(ALPHA) ? fs.readFileSync(ALPHA, 'utf8') : '');
  const inline = alpha ? fromAlpha(alpha) : null;
  if (inline) return inline;
  if (opts.optional) return null;
  throw new Error(
    'the walked city could not be found. Looked for slices/BOHEMIA_CITY_WORLD.html ' +
    'and for an inline CITY_B64 in the alpha. If the world moved again, this ' +
    'file (gates/bohemia_city_src.js) is the ONE place to teach it the new address.');
}

citySrc.STANDALONE = STANDALONE;
citySrc.where = () => (fs.existsSync(STANDALONE) ? STANDALONE : ALPHA + ' (inline CITY_B64)');
module.exports = citySrc;
